### PR TITLE
ELEC-610: Rest API and client

### PR DESCRIPTION
**Pull request by mishazharov**
Hey guys, I'm opening the PR to add a rest server and client to the telemetry backend. In theory, one telemetry client in the car should send data to another one in the cloud.

There are a few changes that are somewhat integrated, so I decided to open this as one review.

If you have any questions, please let me know.

### CODE REVIEW

**Comment by andyjiin** [comment]
I'm not too familiar with Rest APIs in general, maybe @aaronhktan could take a look

**Comment by anikser** [comment]
Were you able to test this on a raspberry pi? If not, we should do that before this gets merged, ideally this weekend.

File: pkg/db/can_db.go [hunk 1]
```diff
@@ -14,24 +14,21 @@ import (
...
+	createCanTbl := `
       CREATE TABLE IF NOT EXISTS
         can (source INTEGER NOT NULL,
              id INTEGER NOT NULL,
              rtr INTEGER NOT NULL,
              dlc INTEGER NOT NULL,
              timestamp INTEGER NOT NULL,
-             data TEXT NOT NULL);`
-	_, err = db.Exec(createTbl)
+						 data TEXT NOT NULL);`
```
Comment (aaronhktan) [inline pkg/db/can_db.go]: Nit: I think the spacing on this line uses tabs where the rest of the SQL statement uses spaces, should be an easy fix

File: pkg/rest/rest_auth.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		log.Errorf("Failed to create OAuth table " + err.Error())
+	}
+}
+
+// AddTokenToDb Adds a token to the authentication DB table
+func AddTokenToDb(token string, db *sql.DB) {
+	if token != "" {
+		hashed := sha256.Sum256([]byte(token))
+		hashedHex := fmt.Sprintf("%x", hashed)
+		insertQuery := "INSERT INTO auth(token) VALUES (?)"
```
Comment (aaronhktan) [inline pkg/rest/rest_auth.go]: Nit: Use multiline string for consistency

File: pkg/cli/start.go [hunk 3]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
+	if dbConnectString != "" {
+
+		db, err = sql.Open(dbDriver, dbConnectString)
+
+		if err != nil {
+			log.Errorf("Failed to open db " + err.Error())
+			return err
+		}
+
+		rest.InitAuthDb(db)
```
Comment (aaronhktan) [inline pkg/cli/start.go]: If an error occurs when creating the authentication table, should we still try to add the token? How necessary is this table to the functioning of the REST API, and if it's important, should we do more than log an error if it doesn't work?

**Review by aaronhktan — COMMENTED** [review]
Thanks for doing this Misha! I'm not super experienced with REST APIs and servers, but I don't see anything egregiously wrong with this PR.

Like @anikser mentioned, has this been tested? Once it's tested and confirmed working, merging would be good!

**Comment by mishazharov** [comment]
Thanks for the feedback everyone, I'll try to address everything soon. This has been tested on x86, I'm hoping to test this on the raspi sometime soon.

**Comment by mishazharov** [comment]
I thought about this PR some more, and that lead me to question whether it is actually useful for FSGP (or at all).

Essentially it just allows the chase car to see the driver display, which doesn't have a lot of information on it to begin with. A good solution would be to display more information (driver display frontend rewrite), but I don't think that's going to be done before FSGP. As well, since there were some talks of scrapping the current golang backend in favour of a python based framework with cantools, there won't be an opportunity to iterate on this design in the current implementation.

Finally, we'd need to host the telemetry backend on some cloud server which would take some initiative.

What do you guys think? Will this still be a useful feature that will benefit us in FSGP, and should I continue to merge this?

File: pkg/cli/start.go [hunk 4]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
 	Start the telemetry server, which will start listening for data on
 	the input device specified via the --source flags
 	`,
-	Example: `  telemetry start --schema=/foo.asciipb [--port=port] [--fake] [--tty=/dev/...] [--db=a.db]`,
-	RunE:    runStart,
+	Example: `
+	telemetry start --schema=/foo.asciipb
+									[--port=port]
+									[--source=(s|f|r)]
```
Comment (aaronhktan) [inline pkg/cli/start.go]: Since the command line arguments are changing, could you update the README (under telemetry-server setup) to use `--source=f` instead of `-f`?

**Comment by aaronhktan** [comment]
@mishazharov Here's what I think:

- I think these changes will be useful, even if we don't get a new telemetry frontend before FSGP. It has some QoL enhancements (typescript in driver display checking the right websocket lol) and it's useful to build on.
- We have @g-jessmuir who'll be taking a look at building a more informative and full-featured front-end.
- The Python-based cantools backend isn't happening before FSGP. I don't think that should be a huge factor to consider right now.

So all in all, I do think it's worth trying to merge it in before FSGP, even if not everything will be used/build before then.

File: client/src/ts/dd_main.ts [hunk 5]
```diff
@@ -160,7 +160,9 @@ function updateDate(): void {
 // Initializations
 
 let prevDir = 4;
-const ws = new WebSocket('ws://localhost:8080/ws');
+const ws = new WebSocket(
+  'ws://' + window.location.hostname + ':' + window.location.port + '/ws'
+);
```
Comment (GarlandZhang) [inline client/src/ts/dd_main.ts:165]: Can the changes in this file be done in a separate PR? It seems independent from all the main changes this PR is achieving.

File: pkg/cli/start.go [hunk 6]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
+func setupFileServer(r *chi.Mux, endpoint string, filesDir string) {
+	if !strings.HasSuffix(endpoint, "/") {
+			panic("Endpoint must be a folder type")
+	}
+	fileServer := http.StripPrefix(
+		endpoint,
+	  http.TimeoutHandler(http.FileServer(http.Dir(filesDir)), 5 * time.Second, "Failed to serve file"),
+	)
+	log.Infof("Mounting fileserver from %s", filesDir)
+	endpoint += "*"
```
Comment (GarlandZhang) [inline pkg/cli/start.go:82]: It's a bit hard to read this way. Could this be done in a separate method to make it easier to read (e.g.;  normalize_endpoint).

File: pkg/cli/start.go [hunk 7]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
+	startCmd.Flags().String("remoteurl", "", "If CAN data should be sent to a remote server")
+
+	viper.SetEnvPrefix(envPrefix)
+	viper.AutomaticEnv()
+	viper.BindPFlags(startCmd.Flags())
 }
 
-func setupURLRouting(r *chi.Mux, messageBus *pubsub.MessageBus) {
-	r.Get("/ws", ws.ServeHTTP(messageBus, ttyPort, fake))
+func setupFileServer(r *chi.Mux, endpoint string, filesDir string) {
```
Comment (GarlandZhang) [inline pkg/cli/start.go]: r => `router` or something a bit easier to read than a letter.

File: pkg/cli/start.go [hunk 8]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
+
+	r.Get("/ws", ws.ServeHTTP(messageBus))
 	workDir, _ := os.Getwd()
 	filesDir := filepath.Join(workDir, "client", "src")
-	r.FileServer("/", http.Dir(filesDir))
+
+	setupFileServer(r, "/", filesDir)
+
+	// Start the rest api if it was requested
+	if strings.Contains(source, "r") {
```
Comment (GarlandZhang) [inline pkg/cli/start.go:103]: This reads a bit unclear but it seems fine for now. If we repeat this kind of logic to check for these symbols elsewhere, it might be better to abstract to some kind of  Source object.

**Review by GarlandZhang — COMMENTED** [review]
hey Misha, I thought I'd just check out your PR to familiarize with the codebase a bit. Honestly, I'm not knowledgeable enough to know what's going on in this PR but I made some code quality comments to make the review process a bit easier (didn't look through everything though). Share your thoughts if you'd like!

File: pkg/cli/start.go [hunk 9]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
 	`,
-	Example: `  telemetry start --schema=/foo.asciipb [--port=port] [--fake] [--tty=/dev/...] [--db=a.db]`,
-	RunE:    runStart,
+	Example: `
+	telemetry start --schema=/foo.asciipb
+									[--port=port]
+									[--source=(s|f|r)]
+									[--rest]
+									[--tty=/dev/...]
+									[--db=a.db]
```
Comment (aaronhktan) [inline pkg/cli/start.go]: This option is also outdated now, please remove!

File: pkg/cli/start.go [hunk 10]
```diff
@@ -24,58 +32,142 @@ var startCmd = &cobra.Command{
...
 	workDir, _ := os.Getwd()
 	filesDir := filepath.Join(workDir, "client", "src")
-	r.FileServer("/", http.Dir(filesDir))
+
+	setupFileServer(r, "/", filesDir)
+
+	// Start the rest api if it was requested
+	if strings.Contains(source, "r") {
+		if db == nil {
+			return fmt.Errorf("You did not specify a db. A db is required to start the REST server")
```
Comment (aaronhktan) [inline pkg/cli/start.go]: Consider specifying which flag needs to be set in the command line argument (in this case, dbConnectString).

File: client/src/ts/dd_main.ts (hunk 5 above)
Comment (mishazharov) [reply client/src/ts/dd_main.ts:165]: This PR depends on this change since otherwise the client webpage tries to connect to the incorrect address. Also since it's such a small and uncontroversial change we can just get it all in at once and avoid the extra overhead

File: pkg/db/can_db.go (hunk 1 above)
Comment (mishazharov) [reply pkg/db/can_db.go]: Nice catch!

File: pkg/cli/start.go (hunk 6 above)
Comment (mishazharov) [reply pkg/cli/start.go:82]: I've considered refactoring this function but reached the conclusion that it is ok as is. This is because the processing that needs to take place in order to make this URL valid for a fileserver would not be reused in other places.

Besides, it is already in a function and it is performing tasks necessary for that function to succeed.

However I have added some whilespace to make it a bit more readable

File: client/src/ts/dd_main.ts (hunk 5 above)
Comment (GarlandZhang) [reply client/src/ts/dd_main.ts:165]: If the PR depends on it, then this change can go in before this current PR. I agree it's not a big deal but I just wanted to mention that. Overall, awesome job man! :D

File: pkg/cli/start.go (hunk 6 above)
Comment (GarlandZhang) [reply pkg/cli/start.go:82]: In terms of readability, I mean that it isn't clear why the star was being added (I figure its for pattern matching but it took me time to look into what it was doing). This is a nit comment so it's not a big deal but just for future reference.

**Comment by mishazharov** [comment]
This has been tested on x86 and arm:

Server command (x86): ./bin/telemetry start --token="test" --source=r --serverPort="8080" --schema=can_messages.asciipb --dbConnectString test.db

The command above will automatically dump all CAN messages into the sqlite DB.

Client command (arm, raspi at home): nice ./bin/telemetry start --source=f --token="test" --remoteurl="http://192.168.1.109:8080" --schema="can_messages.asciipb" &

Video: https://photos.app.goo.gl/Mogmo9VhQd9z6SteA

Note: this was tested with the "fake" data generator. Testing it with socketcan would be nice, but also a bit more involved. Finally we still need to figure out the cloud host situation.

**Comment by aaronhktan** [comment]
I'm gonna try setting something up in Linode - will report back over the weekend.

Another thing: I tried running the server command on my Mac and it's crashing after a few CAN messages:
https://gist.github.com/aaronhktan/7f3cd2ec48ee1c3dc6c87f1f7023f95c

Thanks for testing this though!
